plugins/aiops/pkg/gateway/metricai.go

```diff
@@ -0,0 +1,365 @@
+package gateway
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	grafanav1alpha1 "github.com/grafana-operator/grafana-operator/v4/api/integreatly/v1alpha1"
+	"github.com/nats-io/nats.go"
+	"github.com/rancher/opni/pkg/resources"
+	metricai "github.com/rancher/opni/plugins/aiops/pkg/apis/metricai"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	serverUrl           = "http://metric-ai-service:8090" // URL of the python metric-ai service
+	jobRunDelimiter     = "="                             // delimiter that splits jobid and suffix, in order to save in natsKV
+	dashboardNamePrefix = "metricai-"
+	timeout             = 10 * time.Second
+)
+
+var dashboardSelector = &metav1.LabelSelector{
+	MatchLabels: map[string]string{
+		resources.AppNameLabel:  "grafana",
+		resources.PartOfLabel:   "opni",
+		resources.InstanceLabel: "opni", // TODO: this should be the name of MonitoringCluster
+	},
+}
+
+type RequestError struct {
+	StatusCode int
+	Err        error
+}
+
+func (r *RequestError) Error() string {
+	return r.Err.Error()
+}
+
+func (p *AIOpsPlugin) CreateGrafanaDashboard(ctx context.Context, jobRunId *metricai.MetricAIId) (*metricai.MetricAIAPIResponse, error) {
+	// dashboardJson is generated in the python service. This function simply create a GrafanaDashboard resource with it and apply it
+	res, err := p.GetJobRunResult(ctx, jobRunId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to Get JobRunRes for metricAI: %v", err)
+	}
+	dashboardJson := string(res.JobRunResultDetails)
+	dashboardName := strings.ToLower(strings.ReplaceAll(res.JobRunId, jobRunDelimiter, ""))
+	grafanaDashboards := []*grafanav1alpha1.GrafanaDashboard{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      dashboardNamePrefix + dashboardName,
+				Namespace: os.Getenv("POD_NAMESPACE"),
+				Labels:    dashboardSelector.MatchLabels,
+			},
+			Spec: grafanav1alpha1.GrafanaDashboardSpec{
+				Json: dashboardJson,
+			},
+		},
+	}
+	for _, dashboard := range grafanaDashboards {
+		err := p.k8sClient.Create(ctx, dashboard)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Error Creating Dashboard for metricAI: %v", err)
```

Comment:
"%v" in Errorf, depending on the error, may cast it to bytes or a pointer, `%s` or `%w` is safer for logging a readable error

Reply:
fixed